sklearn/datasets/_arff_parser.py

```diff
@@ -0,0 +1,501 @@
+        elif any(is_classification):
+            raise ValueError(
+                "Mix of nominal and non-nominal targets is not currently supported"
+            )
+
+        # reshape y back to 1-D array, if there is only 1 target column;
+        # back to None if there are not target columns
+        if y.shape[1] == 1:
+            y = y.reshape((-1,))
+        elif y.shape[1] == 0:
+            y = None
+
+    if output_arrays_type == "pandas":
+        return X, y, frame, None
+    return X, y, None, nominal_attributes
+
+
+def _pandas_arff_parser(
+    gzip_file,
+    output_arrays_type,
+    columns_info_openml,
+    feature_names_to_select,
+    target_names_to_select,
+    infer_casting,
+):
+    """ARFF parser using `pandas.read_csv`.
+
+    The metadata being fetch directly to OpenML, one can skip the metadata from
+    the ARFF file and read the data as a CSV file.
+
+    Parameters
+    ----------
+    gzip_file : GzipFile instance
+        The file compressed to be read.
+
+    output_array_type : {"numpy", "sparse", "pandas"}
+        The type of the arrays that will be returned. The possibilities ara:
+
+        - `"numpy"`: both `X` and `y` will be NumPy arrays;
+        - `"sparse"`: `X` will be sparse matrix and `y` will be a NumPy array;
+        - `"pandas"`: `X` will be a pandas DataFrame and `y` will be either a
+          pandas Series or DataFrame.
+
+    columns_info : dict
+        The information provided by OpenML regarding the columns of the ARFF
+        file.
+
+    feature_names_to_select : list of str
+        A list of the feature names to be selected.
+
+    target_names_to_select : list of str
+        A list of the target names to be selected.
+
+    infer_casting : bool
+        Whether or not to infer the type of the data in each column and thus
+        cast the data to the most appropriate type. Activating this option
+        will be more costly but will solved ambiguities regarding the data
+        types provided by the ARFF metadata.
+
+    Returns
+    -------
+    X : {ndarray, sparse matrix, dataframe}
+        The data matrix.
+
+    y : {ndarray, dataframe, series}
+        The target.
+
+    frame : dataframe or None
+        A dataframe containing both `X` and `y`. `None` if
+        `output_array_type != "pandas"`.
+
+    nominal_attributes : list of str or None
+        The names of the features that are categorical. `None` if
+        `output_array_type == "pandas"`.
+    """
+    import pandas as pd
+
+    # read the file until the data section
+    for line in gzip_file:
+        if line.decode("utf-8").lower().startswith("@data"):
+            break
+    # potentially there are some comments before the data section
+    # let's peek a ahead such that the pandas parser does not fail
+    while True:
+        stream_ahead = gzip_file.peek(1)
+        if stream_ahead.decode("utf-8").startswith("%"):
+            line = gzip_file.readline()
+        else:
+            break
```

Comment:
Due to this quirk, we are now obliged to compute the MD5 in advance to not update the hash in both `read1` and `__next__`. In the end, this is not really costly compared to the cost of parsing the CSV file. In addition, it makes the code more readable (not subclassing `GzipFile`).